File: modules/sql/app/model/SqlScenarioReader.java
package model;

import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.LinkedList;
import java.util.List;

import org.apache.ibatis.jdbc.ScriptRunner;

import com.fasterxml.jackson.databind.JsonNode;

import model.exercise.SqlExercise;
import model.exercise.SqlScenario;
import model.exercisereading.ExerciseCollectionReader;
import play.Logger;
import play.data.DynamicForm;
import play.db.Database;

public class SqlScenarioReader extends ExerciseCollectionReader<SqlExercise, SqlScenario> {

  private static final String CREATE_DUMMY = "CREATE DATABASE IF NOT EXISTS ";

  private final Database sqlSelect;
  private final Database sqlOther;

  public SqlScenarioReader(Database theSqlSelect, Database theSqlOther) {
    super("sql", SqlScenario.finder, SqlScenario[].class);
    sqlSelect = theSqlSelect;
    sqlOther = theSqlOther;
  }

  private static void createDatabase(String databaseName, Connection connection) {
    try(Statement createStatement = connection.createStatement()) {
      createStatement.executeUpdate(CREATE_DUMMY + databaseName); // NOSONAR
      connection.setCatalog(databaseName);
    } catch (final SQLException e) {
      Logger.error("There has been an error running an sql script: \"" + CREATE_DUMMY + databaseName + "\"", e);
    }
  }

  public void initRemainingExFromForm(SqlScenario exercise, DynamicForm form) {
    exercise.shortName = form.get(StringConsts.SHORTNAME_NAME);
    exercise.scriptFile = form.get(StringConsts.SCRIPTFILE_NAME);
  }

  @Override
  public SqlScenario instantiate(int id) {
    return new SqlScenario(id);
  }

  public void runCreateScript(Database database, SqlScenario scenario) {
    System.out.println(exerciseType());
    System.out.println(scenario.scriptFile);
    final Path scriptFilePath = Paths.get("conf", "resources", exerciseType(), scenario.scriptFile).toAbsolutePath();

    if(!scriptFilePath.toFile().exists()) {
      Logger.error("File " + scriptFilePath + " does not exist");
      return;
    }

    try {
      final Connection connection = database.getConnection();

      createDatabase(scenario.shortName, connection);

      final ScriptRunner runner = new ScriptRunner(connection);
      runner.setLogWriter(null);
      runner.runScript(new FileReader(scriptFilePath.toFile()));

      connection.close();
    } catch (IOException | SQLException e) {
      Logger.error("Error while executing script file " + scriptFilePath.toString(), e);
    }

  }

  @Override
  public void save(SqlScenario scenario) {
    scenario.save();

    runCreateScript(sqlSelect, scenario);
    runCreateScript(sqlOther, scenario);

    // scenario.getExercises().forEach(delegateReader::save);
  }

  @Override
  public void updateCollection(SqlScenario exercise, JsonNode exerciseNode) {
    exercise.shortName = exerciseNode.get(StringConsts.SHORTNAME_NAME).asText();
    exercise.scriptFile = exerciseNode.get(StringConsts.SCRIPTFILE_NAME).asText();

    List<SqlExercise> exes = new LinkedList<>();
    JsonNode exesNode = exerciseNode.get(StringConsts.EXERCISES_NAME);
    for(JsonNode exNode: exesNode.iterator())
      exes.add(SqlExerciseReader.getInstance().read(exNode));

    List<SqlExercise> exercises = exes;
  }

}

File: modules/sql/app/model/SqlScenarioReader.scala
package model

import java.io.FileReader
import java.nio.file.Paths
import java.sql.{Connection, SQLException}

import com.fasterxml.jackson.databind.JsonNode
import model.CommonUtils.cleanly
import model.exercise.{SqlExercise, SqlScenario}
import model.exercisereading.{ExerciseCollectionReader, ExerciseReader}
import org.apache.ibatis.jdbc.ScriptRunner
import play.Logger
import play.db.Database

import scala.collection.JavaConverters._

class SqlScenarioReader(sqlSelect: Database, sqlOther: Database)
  extends ExerciseCollectionReader[SqlExercise, SqlScenario]("sql", SqlScenario.finder, classOf[Array[SqlScenario]]) {

  val createDummy = "CREATE DATABASE IF NOT EXISTS "

  def createDatabase(databaseName: String, connection: Connection) {
    CommonUtils.cleanly(connection.createStatement)(_.close)(createStatement => {
      try {
        createStatement.executeUpdate(createDummy + databaseName)
        connection.setCatalog(databaseName)
      } catch {
        case e: SQLException => Logger.error(s"""There has been an error running an sql script: "$createDummy $databaseName"""", e)
      }
    })
  }

//  def initRemainingExFromForm(exercise: SqlScenario, form: DynamicForm) {
  //    exercise.shortName = form.get(StringConsts.SHORTNAME_NAME)
  //    exercise.scriptFile = form.get(StringConsts.SCRIPTFILE_NAME)
  //  }

  override def instantiate(id: Int) = new SqlScenario(id)

  def runCreateScript(database: Database, scenario: SqlScenario) {
    println(exerciseType)
    println(scenario.scriptFile)
    val scriptFilePath = Paths.get("conf", "resources", exerciseType, scenario.scriptFile).toAbsolutePath

    if (scriptFilePath.toFile.exists)
      Logger.error("File " + scriptFilePath + " does not exist")
    else {
      cleanly(database.getConnection)(_.close)(connection => {
        val connection = database.getConnection

        createDatabase(scenario.shortName, connection)

        val runner = new ScriptRunner(connection)
        runner.setLogWriter(null)
        runner.runScript(new FileReader(scriptFilePath.toFile))

        connection.close()
      })
    }

  }

  override def save(scenario: SqlScenario) {
    scenario.save()

    runCreateScript(sqlSelect, scenario)
    runCreateScript(sqlOther, scenario)

    // scenario.getExercises.forEach(delegateReader::save)
  }

  override def updateCollection(exercise: SqlScenario, exerciseNode: JsonNode) {
    exercise.shortName = exerciseNode.get(StringConsts.SHORTNAME_NAME).asText
    exercise.scriptFile = exerciseNode.get(StringConsts.SCRIPTFILE_NAME).asText

    exercise.exercises = ExerciseReader.readArray(exerciseNode.get(StringConsts.EXERCISES_NAME),
      SqlExerciseReader.read).map(_.read).asJava
  }

}
